crates/services/p2p/src/discovery/dnsaddr_resolution.rs

```diff
@@ -0,0 +1,101 @@
+use anyhow::anyhow;
+use hickory_resolver::TokioAsyncResolver;
+use libp2p::Multiaddr;
+use std::pin::Pin;
+
+/// The prefix for `dnsaddr` protocol TXT record lookups.
+const DNSADDR_PREFIX: &str = "_dnsaddr.";
+/// The maximum number of DNS lookups when dialing.
+/// This limit is for preventing malicious or misconfigured DNS records from causing infinite recursion.
+const MAX_DNS_LOOKUPS: usize = 10;
+
+#[async_trait::async_trait]
+pub trait DnsLookup {
+    async fn lookup_dnsaddr(&self, addr: &str) -> anyhow::Result<Vec<Multiaddr>>;
+    async fn new() -> anyhow::Result<Box<Self>>;
+}
+
+pub struct DnsResolver {
+    resolver: TokioAsyncResolver,
+}
+
+#[async_trait::async_trait]
+impl DnsLookup for DnsResolver {
+    async fn lookup_dnsaddr(&self, addr: &str) -> anyhow::Result<Vec<Multiaddr>> {
+        self.resolve_recursive(addr, 0).await
+    }
+
+    async fn new() -> anyhow::Result<Box<Self>> {
+        let resolver = TokioAsyncResolver::tokio_from_system_conf()?;
+        Ok(Box::new(Self { resolver }))
+    }
+}
+
+impl DnsResolver {
+    /// Internal method to handle recursive DNS lookups.
+    fn resolve_recursive<'a>(
+        &'a self,
+        addr: &'a str,
+        depth: usize,
+    ) -> Pin<
+        Box<dyn std::future::Future<Output = anyhow::Result<Vec<Multiaddr>>> + Send + 'a>,
+    > {
```

Comment:
Do we need to do the `Pin<Box<...>>` here? That feels like an artifact of `async_trait` which should preferably be removed (as per my other suggestion), or if you want to keep the `async_trait`, this should happen in the trait implementation and not in this private helper method.

Reply:
recursion needs us to pin the future

Reply:
Ah right, makes sense :+1: